File: hashlife.py
from collections import namedtuple
from functools import lru_cache
from collections import Counter

def baseline_life(pts):    
    """The baseline implementation of the Game of Life. Takes a list/set 
    of (x,y) on cells, and returns a new set of on cells in the next
    generation"""
    ns = Counter([(x+a, y+b) for x,y in pts for a in [-1,0,1] for b in [-1,0,1]])
    return Counter([p for p in ns if ns[p]==3 or (ns[p]==4 and p in pts)])


# The base quadtree node
# `k` is the level of the node
# `a, b, c, d` are the children of this node (or None, if `k=0`).
# `n` is the number of on cells in this node (useful for bookkeeping and display) 
# `hash` is a precomputed hash of this node 
# (if we don't do this, Python will recursively compute the hash every time it is needed!) 
_Node = namedtuple("Node", ["k", "a", "b", "c", "d", "n", "hash"])

class Node(_Node):
    def __hash__(self):
        return self.hash

    def __repr__(self):
        return f"Node k={self.k}, {1<<self.k} x {1<<self.k}, population {self.n}"


# base level binary nodes
on = Node(0, None, None, None, None, 1, 1)
off = Node(0, None, None, None, None, 0, 0)

mask = (1 << 63) - 1

@lru_cache(maxsize=2 ** 24)
def join(a, b, c, d):
    """
    Combine four children at level `k-1` to a new node at level `k`.
    If this is cached, return the cached node. 
    Otherwise, create a new node, and add it to the cache.
    """
    n = a.n + b.n + c.n + d.n
    nhash = (
        a.k
        + 2
        + 5131830419411 * a.hash
        + 3758991985019 * b.hash
        + 8973110871315 * c.hash
        + 4318490180473 * d.hash
    ) & mask
    return Node(a.k + 1, a, b, c, d, n, nhash)


@lru_cache(maxsize=1024)
def get_zero(k):
    """Return an empty node at level `k`."""
    if k > 0:
        return join(get_zero(k - 1), get_zero(k - 1), get_zero(k - 1), get_zero(k - 1))
    else:
        return off


def centre(m):
    """Return a node at level `k+1`, which is centered on the given quadtree node."""
    z = get_zero(m.a.k)  # get the right-sized zero node
    return join(
        join(z, z, z, m.a), join(z, z, m.b, z), join(z, m.c, z, z), join(m.d, z, z, z)
    )


def life(a, b, c, d, E, f, g, h, i):
    """The standard life rule, taking eight neighbours and a centre cell E.
    Returns on if should be on, and off otherwise."""
    outer = sum([t.n for t in [a, b, c, d, f, g, h, i]])
    return on if (E.n and outer == 2) or outer == 3 else off


def life_4x4(m):
    """
    Return the next generation of a $k=2$ (i.e. 4x4) cell. 
    To terminate the recursion, at the base level, 
    if we have a $k=2$ 4x4 block, 
    we can compute the 2x2 central successor by iterating over all 
    the 3x3 sub-neighbourhoods of 1x1 cells using the standard Life rule.
    """
    na = life(m.a.a, m.a.b, m.b.a, m.a.c, m.a.d, m.b.c, m.c.a, m.c.b, m.d.a)  # AD
    nb = life(m.a.b, m.b.a, m.b.b, m.a.d, m.b.c, m.b.d, m.c.b, m.d.a, m.d.b)  # BC
    nc = life(m.a.c, m.a.d, m.b.c, m.c.a, m.c.b, m.d.a, m.c.c, m.c.d, m.d.c)  # CB
    nd = life(m.a.d, m.b.c, m.b.d, m.c.b, m.d.a, m.d.b, m.c.d, m.d.c, m.d.d)  # DA
    return join(na, nb, nc, nd)


@lru_cache(maxsize=2 ** 24)
def successor(m, j=None):
    """
    Return the 2**k-1 x 2**k-1 successor, 2**j generations in the future, 
    where j <= k - 2, caching the result.
    """
    if m.n == 0:  # empty
        return m.a
    elif m.k == 2:  # base case
        s = life_4x4(m)
    else:
        j = m.k - 2 if j is None else min(j, m.k - 2)
        c1 = successor(join(m.a.a, m.a.b, m.a.c, m.a.d), j)
        c2 = successor(join(m.a.b, m.b.a, m.a.d, m.b.c), j)
        c3 = successor(join(m.b.a, m.b.b, m.b.c, m.b.d), j)
        c4 = successor(join(m.a.c, m.a.d, m.c.a, m.c.b), j)
        c5 = successor(join(m.a.d, m.b.c, m.c.b, m.d.a), j)
        c6 = successor(join(m.b.c, m.b.d, m.d.a, m.d.b), j)
        c7 = successor(join(m.c.a, m.c.b, m.c.c, m.c.d), j)
        c8 = successor(join(m.c.b, m.d.a, m.c.d, m.d.c), j)
        c9 = successor(join(m.d.a, m.d.b, m.d.c, m.d.d), j)

        if j < m.k - 2:
            s = join(
                (join(c1.d, c2.c, c4.b, c5.a)),
                (join(c2.d, c3.c, c5.b, c6.a)),
                (join(c4.d, c5.c, c7.b, c8.a)),
                (join(c5.d, c6.c, c8.b, c9.a)),
            )
        else:
            s = join(
                successor(join(c1, c2, c4, c5), j),
                successor(join(c2, c3, c5, c6), j),
                successor(join(c4, c5, c7, c8), j),
                successor(join(c5, c6, c8, c9), j),
            )
    return s


def construct(pts):
    """
    Turn a list of (x,y) coordinates into a quadtree
    and return the top-level Node.
    """
    # Force start at (0,0)
    min_x = min(*[x for x, y in pts])
    min_y = min(*[y for x, y in pts])
    pattern = {(x - min_x, y - min_y): on for x, y in pts}
    k = 0
    while len(pattern) != 1:
        # bottom-up construction
        next_level = {}
        z = get_zero(k)
        while len(pattern) > 0:
            x, y = next(iter(pattern))
            x, y = x - (x & 1), y - (y & 1)
            # read all 2x2 neighbours, removing from those to work through
            # at least one of these must exist by definition
            a = pattern.pop((x, y), z)
            b = pattern.pop((x + 1, y), z)
            c = pattern.pop((x, y + 1), z)
            d = pattern.pop((x + 1, y + 1), z)
            next_level[x >> 1, y >> 1] = join(a, b, c, d)
        # merge at the next level
        pattern = next_level
        k += 1
    return pattern.popitem()[1]


def expand(node, x=0, y=0, clip=None, level=0):
    """Turn a quadtree a list of (x,y,gray) triples 
    in the rectangle (x,y) -> (clip[0], clip[1]) (if clip is not-None).    
    If `level` is given, quadtree elements at the given level are given 
    as a grayscale level 0.0->1.0,  "zooming out" the display.
    """

    if node.n == 0:  # quick zero check
        return []
    size = 2 ** node.k
    # bounds check
    if clip is not None:
        if x + size < clip[0] or x > clip[1] or y + size < clip[2] or y > clip[3]:
            return []
    if node.k == level:
        # base case: return the gray level of this node
        gray = node.n / (size ** 2)
        return [(x >> level, y >> level, gray)]
    else:
        # return all points contained inside this node
        offset = size >> 1
        return (
            expand(node.a, x, y, clip, level)
            + expand(node.b, x + offset, y, clip, level)
            + expand(node.c, x, y + offset, clip, level)
            + expand(node.d, x + offset, y + offset, clip, level)
        )


def print_node(node):
    """
    Print out a node, fully expanded    
    """
    points = expand(crop(node))
    px, py = 0, 0
    for x, y, gray in sorted(points, key=lambda x: (x[1], x[0])):
        while y > py:
            print()
            py += 1
            px = 0
        while x > px:
            print(" ", end="")
            px += 1
        print("*", end="")


def is_padded(node):
    """
    True if the pattern is surrounded by at least one sub-sub-block of
    empty space.
    """
    return (
        node.a.n == node.a.d.d.n
        and node.b.n == node.b.c.c.n
        and node.c.n == node.c.b.b.n
        and node.d.n == node.d.a.a.n
    )


def inner(node):
    """
    Return the central portion of a node -- the inverse operation
    of centre()
    """
    return join(node.a.d, node.b.c, node.c.b, node.d.a)


def crop(node):
    """
    Repeatedly take the inner node, until all padding is removed.
    """
    if node.k <= 3 or not is_padded(node):
        return node
    else:
        return crop(inner(node))


def pad(node):
    """
    Repeatedly centre a node, until it is fully padded.
    """
    if node.k <= 3 or not is_padded(node):
        return pad(centre(node))
    else:
        return node


def ffwd(node, n):
    """Advance as quickly as possible, taking n 
    giant leaps"""
    gens = 0
    for i in range(n):
        node = pad(node)
        gens += 1 << (node.k - 2)
        node = successor(node)
    return node, gens


def advance(node, n):
    """Advance node by exactly n generations, using
    the binary expansion of n to find the correct successors"""
    if n == 0:
        return node
    bits = []
    # get the binary expansion, and pad sufficiently
    while n > 0:
        bits.append(n & 1)
        n = n >> 1
        node = centre(node)

    # apply the successor rule
    for k, bit in enumerate(reversed(bits)):
        j = len(bits) - k - 1
        if bit:
            node = successor(node, j)
    return crop(node)


if __name__ == "__main__":
    from lifeparsers import autoguess_life_file
    import time

    # run the breeder forward many generations
    def load_lif(fname):
        pat, comments = autoguess_life_file(fname)
        return construct(pat)

    init_t = time.perf_counter()
    print(ffwd(load_lif("lifep/breeder.lif"), 64))
    t = time.perf_counter() - init_t
    print(f"Computation took {t*1000.0:.1f}ms")
    print(successor.cache_info())
    print(join.cache_info())

    from render import render_img
    ## test the Gosper glider gun
    pat = load_lif("lifep/gun30.LIF")
    pat = load_lif("lifep/gun30.lif")
    render_img(expand(pat))
    plt.savefig("imgs/gun30_0.png", bbox_inches="tight")
    render_img(expand(advance(centre(centre(pat)), 30)))
    plt.savefig("imgs/gun120_30.png", bbox_inches="tight")

    render_img(expand(advance(centre(centre(pat)), 120), level=0))
    plt.savefig("imgs/gun30_120_0.png", bbox_inches="tight")
    render_img(expand(advance(centre(centre(pat)), 120), level=1))
    plt.savefig("imgs/gun30_120_1.png", bbox_inches="tight")
    render_img(expand(advance(centre(centre(pat)), 120), level=2))
    plt.savefig("imgs/gun30_120_2.png", bbox_inches="tight")
    render_img(expand(advance(centre(centre(pat)), 120), level=3))
    plt.savefig("imgs/gun30_120_3.png", bbox_inches="tight")

File: hashlife.c
/*
    Hashlife.

    Design:
    - All nodes are referenced by an opaque identifier.
    - There is an *intern index* that maps from identifiers to pointers to nodes.
    - It matches on (a, b, c, d) to ensure uniqueness.
    - Nodes are reference counted.

*/
#include "hashlife.h"


/* SplitMix64 mixing function */
uint64_t mix64(uint64_t x)
{
    x += 0x9e3779b97f4a7c15ULL;
    x = (x ^ (x >> 30)) * 0xbf58476d1ce4e5b9ULL;
    x = (x ^ (x >> 27)) * 0x94d049bb133111ebULL;
    return x ^ (x >> 31);
}

/* Hash four values into one */
uint64_t hash_quad(const uint64_t a, const uint64_t b, const uint64_t c, const uint64_t d)
{
    uint64_t h = 0x243f6a8885a308d3ULL;
    h ^= mix64(a + 0x9e3779b97f4a7c15ULL * 1);
    h = (h * 0x9e3779b97f4a7c15ULL) ^ (h >> 32);
    h ^= mix64(b + 0x9e3779b97f4a7c15ULL * 2);
    h = (h * 0x9e3779b97f4a7c15ULL) ^ (h >> 32);
    h ^= mix64(c + 0x9e3779b97f4a7c15ULL * 3);
    h = (h * 0x9e3779b97f4a7c15ULL) ^ (h >> 32);
    h ^= mix64(d + 0x9e3779b97f4a7c15ULL * 4);
    h = (h * 0x9e3779b97f4a7c15ULL) ^ (h >> 32);
    return mix64(h);
}

node_id lookup_next(node_table *table, node_id from, uint64_t j)
{
    uint64_t mask = (table->size) - 1;
    uint64_t hash = hash_quad(from, j, from, j);
    node *n = &table->index[hash & mask];
    if(n->from==from && n->j==j)
        return n->to; 
    return UNUSED;
}

void cache_next(node_table *table, node_id from, node_id to, uint64_t j)
{
    uint64_t mask = (table->size) - 1;
    uint64_t hash = hash_quad(from, j, from, j);
    node *n = &table->index[hash & mask];    
    n->from = from;
    n->to = to;
    n->j = j;    
}


/* Centre a node, by surrounding it with zeros of the same size.
    Returns the new node.
 */
node_id centre(node_table *table, node_id m_h)
{
    node_id z;
    node m = *lookup(table, m_h);
    
    z = get_zero(table, m.level - 1);
    node_id a = join(table, z, z, z, m.a);
    node_id b = join(table, z, z, m.b, z);
    node_id c = join(table, z, m.c, z, z);
    node_id d = join(table, m.d, z, z, z);
    return join(table, a, b, c, d);
}

/* Return the zero node of the given size */
node_id get_zero(node_table *table, uint64_t k)
{
    node_id z = OFF;
    for(uint64_t i=0; i<k; i++)
        z = join(table, z, z, z, z);        
    return z;        
}


node *lookup(node_table *table, node_id id)
{
    uint64_t mask = table->size - 1;
    uint64_t index = id & mask;
    node *n = &table->index[index];
    while (n->id != UNUSED && n->id != id)
    {
        // linear probing
        index = (index + 1) & mask;
        n = &table->index[index];    
    }    
    
    return n;
}



/* Double the size of the table, reinserting nodes */
void resize_table(node_table *table)
{
    uint64_t new_size = table->size * 2;
    node *new_index = (node *)calloc(new_size, sizeof(node));
    uint64_t new_mask = new_size - 1;
    uint64_t new_index_pos;
    node *old_n;
    for (uint64_t i = 0; i < table->size; i++)
    {
        old_n = &table->index[i];
        // perform re-insertion of all nodes
        if (old_n->id != UNUSED)
        {
            new_index_pos = old_n->id & new_mask;
            while (new_index[new_index_pos].id != 0)
                new_index_pos = (new_index_pos + 1) & new_mask;
            new_index[new_index_pos] = *old_n;
        }
    }
    free(table->index);
    table->index = new_index;
    table->size = new_size;
}


/* Join four nodes.
   - If the node exists, return it from the table.
   - Otherwise, intern it.
   - Increment the reference counts of the child nodes.
   - Returns the hash of the joined node.
*/

node_id join(node_table *table, node_id a_hash, node_id b_hash, node_id c_hash, node_id d_hash)
{
    uint64_t hash = hash_quad(a_hash, b_hash, c_hash, d_hash);
    node *n = lookup(table, hash);    
    while (n->id != UNUSED)
    {
        if (n->a == a_hash && n->b == b_hash && n->c == c_hash && n->d == d_hash)
            return n->id;   // found it
        hash = mix64(hash); // doppleganger, create a new unique id
        n = lookup(table, hash);
    }

    // not found, so create it
    n->id = hash;
    
    // set children
    n->a = a_hash;
    n->b = b_hash;
    n->c = c_hash;
    n->d = d_hash;
    // set population
    node *a = lookup(table, a_hash);
    node *b = lookup(table, b_hash);
    node *c = lookup(table, c_hash);
    node *d = lookup(table, d_hash);
    n->level = a->level + 1;
    n->pop = a->pop + b->pop + c->pop + d->pop;
    table->count++;

    // resize the table if necessary
    if (table->count * 4 >= table->size)
       resize_table(table);
    return hash;
}

static inline node_id sucjoin(node_table *table, node_id a, node_id b, node_id c, node_id d, uint64_t j)
{
    return successor(table, join(table, a, b, c, d), j);
}

/* Find the successor of the given node, 2^level-2 steps in the future */
node_id successor(node_table *table, node_id id, uint64_t j)
{
    node *n = lookup(table, id);
    uint64_t level = n->level;    
    
    if (j == 0 || j >= level - 2)
        j = level - 2;        
    if (n->pop == 0) // empty
        return n->a;
    if (level == 2) // base case
        return life_4x4(table, id);    
    node_id next = lookup_next(table, id, j);
    if(next != UNUSED) // cached
        return next;    

    node_id c1, c2, c3, c4, c5, c6, c7, c8, c9;
    // copy the actual nodes to prevent changes during lookups
    node a = *lookup(table, n->a);
    node b = *lookup(table, n->b);
    node c = *lookup(table, n->c);
    node d = *lookup(table, n->d);

    c1 = successor(table, a.id, j);  // same as sucjoin(table, a.a, a.b, a.c, a.d, j);
    c2 = sucjoin(table, a.b, b.a, a.d, b.c, j);
    c3 = successor(table, b.id, j); 
    c4 = sucjoin(table, a.c, a.d, c.a, c.b, j);
    c5 = sucjoin(table, a.d, b.c, c.b, d.a, j);
    c6 = sucjoin(table, b.c, b.d, d.a, d.b, j);
    c7 = successor(table, c.id, j); 
    c8 = sucjoin(table, c.b, d.a, c.d, d.c, j);
    c9 = successor(table, d.id, j); 

    /* Not the natural successor; combine parts */
    if (j < level - 2)
    {
        node c1n = *lookup(table, c1);
        node c2n = *lookup(table, c2);
        node c3n = *lookup(table, c3);
        node c4n = *lookup(table, c4);
        node c5n = *lookup(table, c5);
        node c6n = *lookup(table, c6);
        node c7n = *lookup(table, c7);
        node c8n = *lookup(table, c8);
        node c9n = *lookup(table, c9);

        next = join(table,
                    join(table, c1n.d, c2n.c, c4n.b, c5n.a),
                    join(table, c2n.d, c3n.c, c5n.b, c6n.a),
                    join(table, c4n.d, c5n.c, c7n.b, c8n.a),
                    join(table, c5n.d, c6n.c, c8n.b, c9n.a));
        cache_next(table, id, next, j);
        return next;
    }
    else
    {
        /* Natural successor */
        next = join(table,
                            sucjoin(table, c1, c2, c4, c5, j),
                            sucjoin(table, c2, c3, c5, c6, j),
                            sucjoin(table, c4, c5, c7, c8, j),
                            sucjoin(table, c5, c6, c8, c9, j));        
        
        cache_next(table, id, next, j);
        return next;
    }
}


/* Advance time by j steps.
    - Find the MSB of j
    - Make sure that the node has a level at least big enough

*/
node_id advance(node_table *table, node_id id, uint64_t steps)
{
    node *n = lookup(table, id);
    // ensure the node is big enough; n->level must be > log2(steps)+2
    while ((1ULL << (n->level - 2)) < steps)
    {
        // pad the node
        id = centre(table, id);
        n = lookup(table, id);
    }
    // pad to the centre
    id = centre(table, id);
    n = lookup(table, id);
    // advance by 2^j on each set bit j
    for(uint64_t j=1; steps>0; j++, steps>>=1)
        if (steps & 1)
            id = centre(table, successor(table, id, j));
    // crop for the caller
    return crop(table, id);
}

/* Compute the life rule on the 3x3 neighbourhood

    a b c
    d E f
    g h i

    Returns either the hash of the "on" (=2) or "off" (=1) base node.
*/
node_id base_life(node_id a, node_id b, node_id c, node_id d, node_id e, node_id f, node_id g, node_id h, node_id i)
{
    /* hash 1 = off, hash 2 = on, by definition*/    
    int pop_sum = (a == ON) + (b == ON) + (c == ON) + (d == ON) + (f == ON) + (g == ON) + (h == ON) + (i == ON);
    return (pop_sum == 3 || (pop_sum == 2 && e == ON)) ? ON : OFF;
}

node_id life_4x4(node_table *table, node_id id)
{
    node *n = lookup(table, id);    
    node_id na, nb, nc, nd;
    node *a = lookup(table, n->a);
    node *b = lookup(table, n->b);
    node *c = lookup(table, n->c);
    node *d = lookup(table, n->d);
    na = base_life(a->a, a->b, b->a, a->c, a->d, b->c, c->a, c->b, d->a);
    nb = base_life(a->b, b->a, b->b, a->d, b->c, b->d, c->b, d->a, d->b);
    nc = base_life(a->c, a->d, b->c, c->a, c->b, d->a, c->c, c->d, d->c);
    nd = base_life(a->d, b->c, b->d, c->b, d->a, d->b, c->d, d->c, d->d);        
    return join(table, na, nb, nc, nd);    
}

/* Mark every child node, recursively */
void set_flag(node_table *table, node_id id)
{
    node *n = lookup(table, id);
    if(n->level<2)
        return;    
    // set the high bit of level
    n->level = n->level | 0x8000000000000000ULL;
    set_flag(table, n->a);
    set_flag(table, n->b);
    set_flag(table, n->c);
    set_flag(table, n->d);
}

/* Remove all nodes not a child of top */
void vacuum(node_table *table, node_id top)
{
    // walk the tree, marking all reachable nodes
    set_flag(table, top);
    for(uint64_t i=0; i<table->size; i++)
    {
        node *n = &table->index[i];
        if(n->level > 2 && n->id != UNUSED && !(n->level & 0x8000000000000000ULL))
        {
            // unmarked node, delete it
            n->id = UNUSED;
            table->count--;
        }
        n->level = n->level & 0x7FFFFFFFFFFFFFFFULL; // clear the mark bit
    }
}


node_table *create_table(uint64_t initial_size)
{
    node_table *table = (node_table *)malloc(sizeof(node_table));
    if(initial_size < 16)
        initial_size = 16;
    table->size = initial_size;
    table->count = 0;
    table->index = (node *)calloc(table->size, sizeof(node));    
    /* cell level nodes */
    table->index[1] = (node){.level = 0, .a = 0, .b = 0, .c = 0, .d = 0, .pop = 0, .id = OFF}; // off node
    table->index[2] = (node){.level = 0, .a = 0, .b = 0, .c = 0, .d = 0, .pop = 1, .id = ON}; // on node
    table->count = 2;    
    
    return table;
}

/* Return the inner node of half the size in each dimension */
node_id inner(node_table *table, node_id id)
{
    node *n = lookup(table, id);
    node a = *lookup(table, n->a);
    node b = *lookup(table, n->b);
    node c = *lookup(table, n->c);
    node d = *lookup(table, n->d);
    return join(table, a.d, b.c, c.b, d.a);
}

/* Return true if all outer regions are zero (i.e. only inner inset is non-zero)*/
bool is_padded(node_table *table, node_id id)
{
    node *n = lookup(table, id);
    node a = *lookup(table, n->a);
    node b = *lookup(table, n->b);
    node c = *lookup(table, n->c);
    node d = *lookup(table, n->d);
    bool ad = a.pop == lookup(table, a.d)->pop;
    bool bc = b.pop == lookup(table, b.c)->pop;
    bool cb = c.pop == lookup(table, c.b)->pop;
    bool da = d.pop == lookup(table, d.a)->pop;
    return ad && bc && cb && da;
}

/* Repeatedly take the inner node, until no more can be removed */
node_id crop(node_table *table, node_id id)
{
    node *n = lookup(table, id);
    while (n->level > 3 && is_padded(table, id))
    {
        id = inner(table, id);
        n = lookup(table, id);
    }
    return id;
}

/* Repeatedly centre the node, until the node is fully padded */
node_id pad(node_table *table, node_id id)
{
    node *n = lookup(table, id);
    while (n->level < 3 || is_padded(table, id))
    {
        id = centre(table, id);
        n = lookup(table, id);
    }
    return id;
}

/* Set the cell at the given position, returning the new node */
node_id set_cell(node_table *table, node_id id, uint64_t x, uint64_t y, bool state)
{
    node *n = lookup(table, id);
    if (n->level == 0)
    {
        if (state)
            return ON;
        else
            return OFF;
    }

    // expand the node until x < 1<< (level-1) and y < 1 << (level-1)
    while (x >= (1ULL << n->level) || y >= (1ULL << n->level))
    {
        node_id z = get_zero(table, n->level);
        id = join(table, id, z, z, z);
        n = lookup(table, id);
    }

    uint64_t offset = 1 << (n->level - 1);
    node_id a = n->a;
    node_id b = n->b;
    node_id c = n->c;
    node_id d = n->d;
    if (x < offset && y < offset)
        a = set_cell(table, a, x, y, state);
    else if (x >= offset && y < offset)
        b = set_cell(table, b, x - offset, y, state);
    else if (x < offset && y >= offset)
        c = set_cell(table, c, x, y - offset, state);
    else
        d = set_cell(table, d, x - offset, y - offset, state);
    return join(table, a, b, c, d);
}

/* Get the grey level at the given position and level */
float get_cell(node_table *table, node_id id, uint64_t x, uint64_t y, uint64_t level)
{
    node *n = lookup(table, id);
    if (n->level == 0 || n->level == level)
        return n->pop / (float)(1 << (n->level * 2));
    uint64_t size = 1 << n->level;
    // bounds test
    if (x >= size || y >= size)
        return 0.0f;
    // recursive descent
    uint64_t offset = 1 << (n->level - 1);
    if (x < offset && y < offset)
        return get_cell(table, n->a, x, y, level);
    else if (x >= offset && y < offset)
        return get_cell(table, n->b, x - offset, y, level);
    else if (x < offset && y >= offset)
        return get_cell(table, n->c, x, y - offset, level);
    else
        return get_cell(table, n->d, x - offset, y - offset, level);
}
